KIPs/kip-149.md

```diff
@@ -0,0 +1,292 @@
+
+#### Interface of Registry
+
+```solidity
+pragma solidity ^0.8.0;
+
+abstract contract IRegistry {
+    /* ========== VARIABLES ========== */
+    /// The following variables are baked in the interface because their storage layout matters in protocol consensus 
+    /// when inject initial states (system contracts, owner) of the Registry.
+    /// @dev Mapping of system contracts
+    mapping(string => Record[]) public records;
+
+    /// @dev Array of system contract names
+    string[] public names;
+
+    /// @dev Owner of contract
+    address internal _owner;
+
+    /* ========== TYPES ========== */
+    /// @dev Struct of system contracts
+    struct Record {
+        address addr;
+        uint256 activation;
+    }
+
+    /* ========== EVENTS ========== */
+    /// @dev Emitted when the contract owner is updated by `transferOwnership`.
+    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
+    
+    /// @dev Emitted when a new system contract is registered.
+    event Registered(string name, address indexed addr, uint256 indexed activation);
+
+    /* ========== MUTATORS ========== */
+    /// @dev Registers a new system contract.
+    function register(string memory name, address addr, uint256 activation) external virtual;
+    
+    /// @dev Transfers ownership to newOwner.
+    function transferOwnership(address newOwner) external virtual;
+
+    /* ========== GETTERS ========== */
+    /// @dev Returns an address for active system contracts registered as name if exists.
+    ///  It returns a zero address when there's no active system contract with name.
+    function getActiveAddr(string memory name) external virtual returns (address);
+  
+    /// @dev Returns all system contracts registered as name.
+    function getAllRecords(string memory name) external virtual view returns (Record[] memory);
+    
+    /// @dev Returns all names of registered system contracts.
+    function getAllNames() external virtual view returns (string[] memory);
+    
+    /// @dev Returns owner of contract.
+    function owner() external virtual view returns (address);
+}
+```
+
+#### Methods
+
+```solidity
+function register(string memory name, address addr, uint256 activation)
+```
+
+Registers a new system contract. It will be activated at `activation`. It overwrites the predecessor if a predecessor system contract exists and is not yet active. Passing `addr == address(0)` is an implicit deprecation for the `name`, meaning the `name` will no longer be used.
+
+The function validates the following requirements:
+
+- The function caller MUST be a governance contract.
+
+- The function MUST revert if a `name` is an empty string.
+
+- The function MUST revert if `activation < block.number`.
+
```

Comment:
```suggestion

- The function MUST revert if `addr` is in the range of precompiled contract address range

- The function MUST revert if `activation < block.number`.
```

Reply:
1. Why do we need to revert if it's in the range of precompiled contract range? Since we don't know the future usage of precompiled contracts, I think it's better not to prohibit them in KIP explicitly.

2. Good point. We also considered many cases in various conditions but determined the current design since:
    a. System contract registration will be done with much consideration and governance(KIP-81) discussion in the future, and activation will be given with enough buffer if it goes with a hard fork or needs specific block timing.
    b. Emergency activation/deprecation or any methods may be considered a backdoor.

If you think of any needed buffer or emergency functions, it's better to discuss them with other core team members. cc @blukat29 @ian0371 .

Reply:
I'm not exactly sure whether this is needed or not. Based on the communication with @ian0371 and @hyeonLewis, it seems like it is not necessary to include. If any other opinion, please let me know.